### PR TITLE
Stable IPP distanceTransform with single thread

### LINKED ISSUE #24082: Non-deterministic behavior in cv2.distanceTransform()
### System Information

OpenCV python version: 4.5.5
Operating System: Ubuntu 16.04
Python version: 3.9.15

### Detailed description

I am using cv2.distanceTransform in a deep learning application and have noticed non-deterministic training results in pytorch lightning. After setting manual seeds for all random number generators in my setup, the training was still not deterministic so I decided to have a closer look at the used data. I use cv2.distanceTransform in the ground truth data generation in my training and have noticed variation in the output using the same input.

It's possible to mitigate the problem by rounding the results to one decimal, although this is not the best solution ...
```python
result = np.round(cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE), decimals=1)
```
Is there a better way to fix this problem?

### Steps to reproduce

```python
import numpy as np
import cv2

num_tests = 100
shape = (800, 800)
for i in range(num_tests):
    feat_map = np.full(shape=shape, fill_value=255, dtype=np.uint8)
    # Creating a random line
    random_x_start = np.random.randint(0, 399)
    random_x_end = np.random.randint(400, 799)
    random_y_start = np.random.randint(0, 399)
    random_y_end = np.random.randint(400, 799)
    X, Y = np.mgrid[random_x_start:random_x_end:1, random_y_start:random_y_end:1]
    line = np.vstack((X.flatten(), Y.flatten())).transpose().reshape(-1, 1, 2)
    cv2.polylines(feat_map, [line], isClosed=False, color=0)
    result_1 = cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE)
    result_2 = cv2.distanceTransform(feat_map, distanceType=cv2.DIST_L2, maskSize=cv2.DIST_MASK_PRECISE)
    np.testing.assert_array_equal(result_1, result_2)
```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [ ] I updated to the latest OpenCV version and the issue is still there
- [ ] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by dkurt**
### Pull Request Readiness Checklist

resolves #24082

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/imgproc/src/distransform.cpp [hunk 1]
```diff
@@ -796,6 +796,17 @@ void cv::distanceTransform( InputArray _src, OutputArray _dst, OutputArray _labe
...
+                        for (int i = 0; i < dst.rows; ++i)
+                        {
+                            float* row = dst.ptr<float>(i);
+                            for (int j = 0; j < dst.cols; ++j)
+                            {
+                                float rounded = cvRound(row[j]);
+                                if (fabs(row[j] - rounded) <= correctionDiff)
+                                    row[j] = rounded;
+                            }
+                        }
```
Comment (asmorkalov) [inline modules/imgproc/src/distransform.cpp]: I recommend you to add comment with link to the original bug and couple of words why the loop is needed. The reason of "clipping" is not obvious.

**Comment by asmorkalov** [comment]
cc @eplankin

File: modules/imgproc/src/distransform.cpp (hunk 1 above)
Comment (dkurt) [reply modules/imgproc/src/distransform.cpp]: Sure, added the comment

**Comment by asmorkalov** [comment]
The new test fails on MacOS Intel:
```
[ RUN      ] Imgproc_DistanceTransform.ipp_deterministic_corner
/Users/opencv-cn/GHA-OCV-4/_work/opencv/opencv/opencv/modules/imgproc/test/test_distancetransform.cpp:429: Failure
Expected equality of these values:
  expected
    Which is: 2937
  dist.at<float>(0, i)
    Which is: 2937
diff: 4.882812e-04
[  FAILED  ] Imgproc_DistanceTransform.ipp_deterministic_corner (0 ms)
```
